File: 소스코드/랜덤포레스트,의사결정나무.py
#의사결정 나무,랜덤포레스트
import pandas as pd
import numpy as np
from sklearn.metrics import mean_squared_error
from SyncRNG import SyncRNG
import os
os.getcwd()
os.chdir('diabetes.csv')

#데이터로드
raw_data = pd.read_csv('diabetes.csv')


# 데이터 셋  7:3 으로 분할
v=list(range(1,len(raw_data)+1))
s=SyncRNG(seed=42)
ord=s.shuffle(v)
idx=ord[:round(len(raw_data)*0.7)]

# R에서는 데이터프레임이 1부터 시작하기 때문에
# python에서 0행과 R에서 1행이 같은 원리로
# 같은 인덱스 번호를 가진다면 -1을 해주어 같은 데이터를 가지고 오게 한다.
# 인덱스 수정-R이랑 같은 데이터 가져오려고
for i in range(0,len(idx)):
    idx[i]=idx[i]-1

# 학습데이터, 테스트데이터 생성
train=raw_data.loc[idx] # 70%
#train=train.sort_index(ascending=True)
test=raw_data.drop(idx) # 30%

#전처리

y_train=pd.DataFrame(train.Outcome)
x_train=pd.DataFrame(train.drop('Outcome',axis=1))
y_test=pd.DataFrame(test.Outcome)
x_test=pd.DataFrame(test.drop('Outcome',axis=1))

#랜덤 포레스트
from sklearn.ensemble import RandomForestClassifier
random_model = RandomForestClassifier(n_estimators=100,max_features=4,max_depth=5)
random_model.fit(x_train,y_train)
random_pred = random_model.predict(x_test)
#결과
output = pd.DataFrame({'idx': x_test.index, 'Outcome': random_pred})
output.head()
#평가
from sklearn.metrics import classification_report
from sklearn import metrics

print(classification_report(y_test, random_pred))
metrics.f1_score(y_test, random_pred)
#0.6474820143884891

#의사결정 나무
from sklearn.tree import DecisionTreeClassifier
tree_model = DecisionTreeClassifier(max_depth=3,min_samples_leaf=3,min_samples_split=2)
tree_model.fit(x_train, y_train)
tree_pred = tree_model.predict(x_test)

#결과
output = pd.DataFrame({'idx': x_test.index, 'Outcome': tree_pred})
output.head()
#평가
from sklearn.metrics import classification_report
from sklearn import metrics

print(classification_report(y_test, tree_pred))
metrics.f1_score(y_test, tree_pred)

File: 소스코드/랜덤포레스트,의사결정나무.R
# R Baseline
library(SyncRNG) 
library(caret)
library(ModelMetrics)
#데이터 불러오기&전처리
df<-read.csv('diabetes.csv')

#종속변수 factor형으로
df$Outcome<-as.factor(df$Outcome)

#데이터분할
v <- 1:nrow(df)
s <- SyncRNG(seed=42)
idx <- s$shuffle(v)[1:round(nrow(df)*0.7)] 
idx[1:length(idx)]
train_df <- df[idx,] 
test_df <- df[-idx,]

###########랜덤포레스트 모델 train############
#R 모델
library(randomForest)
random_model<-randomForest(Outcome~.,train_df)
random_pred<-predict(random_model,subset(test_df,select=-c(Outcome)))

#caret 기본 모델
caret_random_model <- train(Outcome ~ ., data = train_df, 
                         method = "rf")
caretrandom_pred<-predict(caret_random_model, newdata = test_df)

#caret 튜닝모델
mtry <- sqrt(ncol(train_df))
caretrandomGrid <-  expand.grid(mtry = mtry)
caret_tune_random_model <- train(Outcome ~ ., data = train_df, 
                              method = "rf",
                              tuneGrid= caretrandomGrid)
caret_tune_random_pred<-predict(caret_tune_random_model, newdata = test_df)
################################

#평가
caret::confusionMatrix(random_pred,test_df$Outcome)$byClass #R 평가
caret::confusionMatrix(caretrandom_pred,test_df$Outcome)$byClass #caret평가
caret::confusionMatrix(caret_tune_random_pred,test_df$Outcome)$byClass #caret tune평가

###########의사결정나무 모델 train############
#R 모델
# install.packages("tree")
library(rpart)
tree_model<- rpart(Outcome~., data= train_df)
tree_pred<-predict(tree_model,subset(test_df,select=-c(Outcome)),type='class')
#caret 기본 모델
# fit the model
caret_tree_model = train(Outcome ~ ., 
                  data=train_df, 
                  method="rpart", 
                  )
caret_tree_pred<-predict(caret_tree_model, newdata = test_df)
#caret 튜닝모델
mtry <- sqrt(ncol(train_df))
cp_grid <- data.frame(cp = seq(0.02, .2, .02))
caret_tune_tree_model <- train(Outcome ~ ., data = train_df, 
                                 method = "rpart",
                               trControl=trainControl(method='cv'))
caret_tune_tree_pred<-predict(caret_tune_tree_model, newdata = test_df)
################################

#평가
library(e1071)
treepred <- predict(tree_model,test_df,type="class")
plot(treepred)
confusionMatrix(treepred,test_df$Outcome)

caret::confusionMatrix(tree_pred,test_df$Outcome)$byClass #R 평가
caret::confusionMatrix(caret_tree_pred,test_df$Outcome)$byClass #caret평가
caret::confusionMatrix(caret_tune_tree_pred,test_df$Outcome)$byClass #caret tune평가
